### PR TITLE
Impl integrity test for runtime

### PR DESCRIPTION
**Pull request by thiolliere**
fixes https://github.com/paritytech/substrate/issues/5570

* trait IntegrityTest is created,
* decl_module implement it automatically or using user provided `fn integrity_test`
* construct_runtime create a test that call all of them.
* decl_error test is renamed construct_runtime and one test is added. This is to reuse its structure.

### CODE REVIEW

File: frame/support/test/tests/construct_runtime.rs [hunk 1]
```diff
@@ -137,3 +148,11 @@ fn check_module2_error_type() {
 		Err(DispatchError::Module { index: 2, error: 0, message: Some("Something") }),
 	);
 }
+
+#[test]
+fn integrity_test_works() {
+	// If __construct_runtime_integrity_test::test is executed then the atomic value will be
+	// modified two times in the test suite.
+	__construct_runtime_integrity_test::test();
+	assert!(INTEGRITY_TEST_EXEC.load(Ordering::Relaxed) > 1);
```
Comment (thiolliere) [inline frame/support/test/tests/construct_runtime.rs:157]: cargo test seems to always execute `__construct_runtime_integrity_test::test` before, we could have assert value is 2 but I prefer not to rely on execution order of tests.

File: frame/support/procedural/src/construct_runtime/mod.rs [hunk 2]
```diff
@@ -406,3 +409,17 @@ fn find_system_module<'a>(
...
 }
+
+fn decl_integrity_test(scrate: &TokenStream2) -> TokenStream2 {
+	quote!(
+		#[cfg(test)]
+		mod __construct_runtime_integrity_test {
+			use super::*;
+
+			#[test]
+			pub fn test() {
```
Comment (bkchr) [inline frame/support/procedural/src/construct_runtime/mod.rs:420]:
```suggestion
			pub fn runtime_integrity_tests() {
```

File: frame/support/src/traits.rs [hunk 3]
```diff
@@ -192,6 +192,15 @@ macro_rules! impl_filter_stack {
 	}
 }
 
+/// Type that provide some integrity tests.
+///
+/// This implemented for modules by `decl_module`.
+#[impl_for_tuples(30)]
+pub trait IntegrityTest {
+	/// Run integrity test.
```
Comment (bkchr) [inline frame/support/src/traits.rs:200]:
```suggestion
	/// Run integrity test.
	///
	/// The test is not executed in a externalities provided environment.
```

Maybe we should add this?

File: frame/support/test/tests/construct_runtime.rs (hunk 1 above)
Comment (thiolliere) [reply frame/support/test/tests/construct_runtime.rs]: stimmt 👍

File: frame/support/src/dispatch.rs [hunk 4]
```diff
@@ -269,8 +269,11 @@ impl<T> Parameter for T where T: Codec + EncodeLike + Clone + Eq + fmt::Debug {}
 ///   * `fn on_finalize() -> frame_support::weights::Weight`
 ///
 /// * `offchain_worker`: Executes at the beginning of a block and produces extrinsics for a future block
-/// upon completion. Using this function will implement the
-/// [`OffchainWorker`](./traits/trait.OffchainWorker.html) trait.
+///   upon completion. Using this function will implement the
+///   [`OffchainWorker`](./traits/trait.OffchainWorker.html) trait.
+/// * `integrity_test`: Executes in a test generated by `construct_runtime`, note it doesn't
+///   executes in a externalities provided environment. Implement
```
Comment (gavofyork) [inline frame/support/src/dispatch.rs:275]:
```suggestion
///   execute in an externalities-provided environment. Implement
```

**Pull request merged**